### PR TITLE
Fix generic_parameter_specialization_map in the presence of non-symmetric generic recursion [TG-7815]

### PR DESCRIPTION
**Pull request by NathanJPhillips**
This is a replacement for the closed PR #4913. It adopts a different approach as that approach was shown to be lacking.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- N/A My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/src/java_bytecode/select_pointer_type.cpp [hunk 1]
```diff
@@ -82,26 +81,25 @@ pointer_typet select_pointer_typet::specialize_generics(
...
-          visited_nodes);
-
-        pointer_typet replacement_array_type = java_array_type('a');
-        replacement_array_type.subtype().set(ID_element_type, new_array_type);
-        return replacement_array_type;
-      }
-    }
+    const auto *array_element_type =
+      type_try_dynamic_cast<pointer_typet>(java_array_element_type(*subtype));
+    if(array_element_type == nullptr)
```
Comment (smowton) [inline jbmc/src/java_bytecode/select_pointer_type.cpp:66]: This commit ("Use type_try_dynamic_cast") does more than its message says (it also switches from `if(x) y` to `if(!x) return; y;` style)

File: jbmc/src/java_bytecode/Makefile [hunk 2]
```diff
@@ -5,6 +5,7 @@ SRC = assignments_from_json.cpp \
       ci_lazy_methods_needed.cpp \
       convert_java_nondet.cpp \
       expr2java.cpp \
+      generic_parameter_specialization_map.cpp \
```
Comment (smowton) [inline jbmc/src/java_bytecode/Makefile:8]: Commit message for "Change data storage for generic_parameter_specialization_mapt" has got a bit out of hand -- the "Added ability to output..." bit probably doesn't need to be there if we're adding it at the same time as the map itself, and similarly "Add test..."

File: jbmc/src/java_bytecode/generic_parameter_specialization_map.h [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+{
+  if(map.map.container_to_specializations.empty())
+    stm << "empty map";
+  else
+    stm << "map:\n";
+  std::vector<std::vector<irep_idt>> container_to_params(
+    map.map.container_to_specializations.size());
+  for(const auto &elt : map.map.param_to_container)
+  {
+    std::vector<irep_idt> &params = container_to_params[elt.second.first];
```
Comment (smowton) [inline jbmc/src/java_bytecode/generic_parameter_specialization_map.h]: Using structs instead of `pair` would make it easier to understand what you're referring to here

File: jbmc/src/java_bytecode/generic_parameter_specialization_map_keys.cpp [hunk 4]
```diff
@@ -151,17 +123,18 @@ void generic_parameter_specialization_map_keyst::insert_pairs_for_symbol(
   {
     return;
   }
+
```
Comment (smowton) [inline jbmc/src/java_bytecode/generic_parameter_specialization_map_keys.cpp:126]: Spacing fixups should be in their own commit

File: jbmc/src/java_bytecode/select_pointer_type.cpp [hunk 5]
```diff
@@ -39,47 +36,33 @@ pointer_typet select_pointer_typet::convert_pointer_type(
...
   const generic_parameter_specialization_mapt
-    &generic_parameter_specialization_map,
-  generic_parameter_recursion_trackingt &visited_nodes) const
+    &generic_parameter_specialization_map) const
 {
   auto parameter = type_try_dynamic_cast<java_generic_parametert>(pointer_type);
   if(parameter != nullptr)
   {
-    const irep_idt &parameter_name = parameter->get_name();
+    irep_idt parameter_name = parameter->get_name();
```
Comment (smowton) [inline jbmc/src/java_bytecode/select_pointer_type.cpp:35]: No need for this to change? `const irep_idt &` is pretty widely used for a const `irept`'s data, this probably isn't the place to fight that battle

File: jbmc/src/java_bytecode/generic_parameter_specialization_map_keys.h [hunk 6]
```diff
@@ -23,12 +23,10 @@ class generic_parameter_specialization_map_keyst
 public:
   /// Initialize a generic-parameter-specialization-map entry owner operating
   /// on a given map. Initially it does not own any map entry.
-  /// \param _generic_parameter_specialization_map: map to operate on.
+  /// \param generic_parameter_specialization_map: map to operate on.
```
Comment (smowton) [inline jbmc/src/java_bytecode/generic_parameter_specialization_map_keys.h:26]: Probably split commit "Code cleanup", e.g. "factor out get_final_name_component", one commit per variable rename, the stuff with `new_class_type`, the re-spacing are all pretty independent changes.

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 7]
```diff
@@ -1200,6 +1200,17 @@ void mark_java_implicitly_generic_class_type(
     java_implicitly_generic_class_typet new_class_type(
       class_type, implicit_generic_type_parameters);
 
+    // Prepend existing parameters so choose those above any inherited
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:1208]: This commit should include a test

File: jbmc/src/java_bytecode/select_pointer_type.cpp (hunk 5 above)
Comment (NathanJPhillips) [reply jbmc/src/java_bytecode/select_pointer_type.cpp:35]: I assign to it at the bottom of the loop.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 833e1b70d70e4e75c6bdb9347c043c5d44dedf47).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120497981
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: fcda9ce16368ab0dbf0fcfbe99e4c9872cdf9a54).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120498877
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

File: jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    {
+      const auto param_it = param_to_container.find(parameter.get_name());
+      INVARIANT(
+        param_it != param_to_container.end(),
+        "Some but not all type parameters already mapped");
+      INVARIANT(
+        param_it->second.container_index == container_index,
+        "Not all type parameters assigned to same container");
+      INVARIANT(
+        param_it->second.param_index == param_index++,
```
Comment (smowton) [inline jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp]: Eeep! Let's not have `INVARIANT` conditions with side-effects, there are some compilation settings where the first arg is discarded

File: jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  std::stringstream msg;
+  msg << "mapping <";
+  join_strings(
+    msg,
+    parameters.begin(),
+    parameters.end(),
+    ", ",
+    [](const java_generic_parametert &param) { return param.get_name(); });
+  msg << ">";
+  std::string msg_str = msg.str();
```
Comment (smowton) [inline jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp]: Debugging code accidentally left in?

File: jbmc/unit/java_bytecode/java_bytecode_parse_generics/parse_derived_generic_class.cpp [hunk 10]
```diff
@@ -187,7 +187,9 @@ SCENARIO(
 
     const class_typet &derived_class_type =
       require_type::require_complete_java_implicitly_generic_class(
-        derived_symbol.type, {"java::ContainsInnerClassGeneric::T"});
+        derived_symbol.type,
```
Comment (smowton) [inline jbmc/unit/java_bytecode/java_bytecode_parse_generics/parse_derived_generic_class.cpp:190]: Double-check: do any of these tests exercise triple-nesting?
```
public class A<T> {
  class B {
    class C {
      T field;
    }
  }
}
```
That sort of thing?

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 11]
```diff
@@ -1205,6 +1205,17 @@ void mark_java_implicitly_generic_class_type(
...
+    if(is_java_generic_class_type(class_type))
+    {
+      const java_generic_class_typet::generic_typest &class_type_params =
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:1211]: Commit message for "Make explicit type parameters shadow implicit ones with the same names" -- can you include a brief example of what you mean by precedence here? Are we talking `class A<T> { class B<T> { T field // which T is that? } }`?

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 12]
```diff
@@ -1030,23 +1030,21 @@ static void find_and_replace_parameter(
...
-    PRECONDITION(
-      get_without_final_name_component(parameter_full_name)
-        .compare(get_without_final_name_component(
-          replacement_parameter_full_name)) > 0);
-
-    parameter.type_variable_ref().set_identifier(
-      replacement_parameter_full_name);
-  }
+  // A replacement parameter was found, update the identifier
+  const std::string &replacement_parameter_full_name =
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:1038]: Message: Sp congnative -> cognitive

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 13]
```diff
@@ -1197,19 +1198,21 @@ void mark_java_implicitly_generic_class_type(
...
     }
 
-    for(auto &base : class_type.bases())
+    for(auto &base : new_class_type.bases())
     {
       find_and_replace_parameters(
         base.type(), implicit_generic_type_parameters);
     }
+
+    class_symbol.type = new_class_type;
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:1231]: ! This which was in the Code Cleanup commit suddenly looks like a bugfix -- the results of `find_and_replace_parameters` would previously have been wrongly discarded, right? Can we test the fix?

Comment (NathanJPhillips) [reply jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:1231]: My assumption was that the thing we were changing was a reference to the same and this was just clarifying things. I'll check.

Comment (NathanJPhillips) [reply jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:1231]: It was. I've updated the commit message to make this clear.

File: jbmc/unit/java_bytecode/java_bytecode_parse_generics/parse_derived_generic_class.cpp (hunk 10 above)
Comment (NathanJPhillips) [reply jbmc/unit/java_bytecode/java_bytecode_parse_generics/parse_derived_generic_class.cpp:190]: The other test in that commit does: https://github.com/diffblue/cbmc/commit/aec9e89c82c9042b70342de9b9bedee96c5a99b4#diff-157ff943932394172d454093024a926cR120

**Review by smowton — APPROVED** [review]
Brilliant, thanks for putting so much effort into this one!

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: f5f62cfb45573b1af6a7a15f77a44ccb69269b74).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120739813
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Comment by NathanJPhillips** [comment]
Passing CI for Diffblue bump here: https://github.com/diffblue/test-gen/runs/177289127

File: jbmc/src/java_bytecode/generic_parameter_specialization_map.h [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    {
+    }
+  };
+
+  template <typename ostream>
+  friend ostream &operator<<(ostream &stm, const printert &map);
+};
+
+template <typename ostream>
+ostream &operator<<(
```
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/generic_parameter_specialization_map.h]: :grey_question: Should `ostream` end in a `t` like `ostreamt` if it is a type?

File: jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+  {
+    container_index = container_to_specializations.size();
+    container_to_specializations.emplace_back();
+    std::size_t param_index = 0;
+    for(const java_generic_parametert &parameter : parameters)
+    {
+      const auto result = param_to_container.emplace(
+        parameter.get_name(), container_paramt{container_index, param_index++});
+      INVARIANT(
+        result.second, "Some but not all type parameters already mapped");
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp]: I'm not sure I understand the invariant message; there is no verb. :ox:

File: jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+  else
+  {
+    container_index = first_param_it->second.container_index;
+    std::size_t param_index = 0;
+    for(const java_generic_parametert &parameter : parameters)
+    {
+      const auto param_it = param_to_container.find(parameter.get_name());
+      INVARIANT(
+        param_it != param_to_container.end(),
+        "Some but not all type parameters already mapped");
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp]: :ox:

File: jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+      INVARIANT(
+        param_it->second.container_index == container_index,
+        "Not all type parameters assigned to same container");
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp]: :ox:

File: jbmc/src/java_bytecode/generic_parameter_specialization_map.h [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+#include <vector>
+
+#include <util/string_utils.h>
+
+#include "expr2java.h"
+#include "java_types.h"
+
+class generic_parameter_specialization_mapt
+{
+private:
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/generic_parameter_specialization_map.h:25]: :pick: the public section usually appears before the private one in cbmc

File: jbmc/src/java_bytecode/generic_parameter_specialization_map.h [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+
+#include <stack>
+#include <vector>
+
+#include <util/string_utils.h>
+
+#include "expr2java.h"
+#include "java_types.h"
+
+class generic_parameter_specialization_mapt
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/generic_parameter_specialization_map.h:23]: This needs some documentation

File: jbmc/src/java_bytecode/generic_parameter_specialization_map.h [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+    }
+  };
+
+  template <typename ostream>
+  friend ostream &operator<<(ostream &stm, const printert &map);
+};
+
+template <typename ostream>
+ostream &operator<<(
+  ostream &stm,
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/generic_parameter_specialization_map.h]: :question: Why does `ostream` need to be a template parameter instead of being just `std::ostream`?

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 21]
```diff
@@ -1142,8 +1142,9 @@ void mark_java_implicitly_generic_class_type(
 {
   const std::string qualified_class_name = "java::" + id2string(class_name);
   PRECONDITION(symbol_table.has_symbol(qualified_class_name));
+  // This will have its type changed
```
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:1142]: If you move the call to `get_writeable_ref` into your new `class_symbol.type = ...` line at the bottom, then this comment will no longer be necessary.

**Review by thomasspriggs — APPROVED** [review]
Looks like this fixes some stuff and I can't see any problems. Therefore this has my approval.

File: jbmc/src/java_bytecode/generic_parameter_specialization_map.h (hunk 20 above)
Comment (NathanJPhillips) [reply jbmc/src/java_bytecode/generic_parameter_specialization_map.h]: It could be restricted to `std::ostream` but it doesn't need to be. I've added comment to that effect.

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp (hunk 21 above)
Comment (NathanJPhillips) [reply jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:1142]: Then we would have two lookups though...

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 186adb8aee29a624594d63fdf7ea824f7aa65753).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121391281
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
